utils/geoutils.py

```diff
@@ -80,21 +84,24 @@ def clip_raster_with_gpkg(raster, gpkg, debug=False):
     coords = getFeatures(geo)
 
     # clip the raster with the polygon
-    out_img, out_transform = mask(dataset=raster, shapes=coords, crop=True)
-
-    out_meta = raster.meta.copy()
-    out_meta.update({"driver": "GTiff",
-                     "height": out_img.shape[1],
-                     "width": out_img.shape[2],
-                     "transform": out_transform})
-
-    out_tif = f"{Path(raster.name).stem}_clipped{Path(raster.name).suffix}"
-    with rasterio.open(out_tif, "w", **out_meta) as dest:
-        if debug:
-            logging.debug(f"writing clipped raster to {out_tif}")
-            dest.write(out_img)
-
-    return out_img, dest
+    out_tif = Path(raster.name).parent / f"{Path(raster.name).stem}_clipped{Path(raster.name).suffix}"
+    if os.path.isfile(out_tif):
+        return out_tif
+    else:
+        try:
+            out_img, out_transform = mask(dataset=raster, shapes=coords, crop=True)
+            out_meta = raster.meta.copy()
+            out_meta.update({"driver": "GTiff",
+                             "height": out_img.shape[1],
+                             "width": out_img.shape[2],
+                             "transform": out_transform})
+            with rasterio.open(out_tif, "w", **out_meta) as dest:
+                print(f"writing clipped raster to {out_tif}")
+                dest.write(out_img)
+            return out_tif
+        except ValueError as e:  # if gpkg's extent outside raster: "ValueError: Input shapes do not overlap raster."
+            # TODO: warning or exception? if warning, except must be set in images_to_samples
```

Comment:
does `utils.readers.image_reader_as_array( ... )` still need to call `clip_raster_with_gpkg( ... )` ?
because it still expects 2 returns (not 1)

Reply:
**Warning** (I think)
But this (Lines 102-104) should prob also return something

thinking off the top of my head to fix: I would say 
`return False`
pass thru `process_raster_img` 
after `rst_pth, r_ = process_raster_img(img_pan, gpkg)` test if `rst_path==False` then trigger a flag?
then have the flag `break` out of loop